### PR TITLE
scx_layered: Improve layer core growth algos

### PR DESCRIPTION
**Pull request by hodgesds**
Improve layer core growth algos to be aware of the number of layers. When applying layer core growth stategies offset core iteration order by a chunked layer offset. This makes `Linear`, `Reverse` and other growth algos more useful in that when multiple layers use the same growth strategy they will not longer attempt to grow using the same core order.

Example with changing the default `stress-ng` layer to `Sticky`:
image

### CODE REVIEW

File: scheds/rust/scx_layered/src/layer_core_growth.rs [hunk 1]
```diff
@@ -121,12 +124,28 @@ impl Default for LayerGrowthAlgo {
...
+        self.layer_specs
+            .iter()
+            .position(|spec| spec.name == *self.spec.name)
+            .unwrap_or(0)
+    }
+
+    fn rotate_layer_offset(&self, vec: &'a mut Vec<usize>) -> &Vec<usize> {
+        let offset = self.layer_offset();
+        let num_cores = self.topo.all_cores.len();
+        let chunk = (num_cores as f64 / self.layer_specs.len() as f64).ceil() as usize;
```
Comment (htejun) [inline scheds/rust/scx_layered/src/layer_core_growth.rs:144]: `.div_ceil()`?

**Review by htejun — APPROVED** [review]

File: scheds/rust/scx_layered/src/layer_core_growth.rs [hunk 2]
```diff
@@ -121,12 +124,28 @@ impl Default for LayerGrowthAlgo {
...
     spec: &'a LayerSpec,
     layer_idx: usize,
     topo: &'a Topology,
 }
 
 impl<'a> LayerCoreOrderGenerator<'a> {
+    fn layer_offset(&self) -> usize {
+        self.layer_specs
+            .iter()
+            .position(|spec| spec.name == *self.spec.name)
```
[the remark this answers is not in the document]
Comment (hodgesds) [reply scheds/rust/scx_layered/src/layer_core_growth.rs:137]: Yeah, `layer_idx` is probably better. I wish it was a field on the `LayerSpec`.

**Review by JakeHillion — APPROVED** [review]
Nice! I'm hoping with this we can write some growth algorithms that have good avoidance while also spreading out effectively.

**Comment by htejun** [comment]
BTW, I don't think this is a viable long term direction. This is useful for quickly experimenting with different ordering but this approach can't reliably achieve an outcome that we ultimately want. Imagine a scenario where three layers are changing their sizes dynamically.

Let's say there are 4 llcs with 4 cpus for simplicity. Let's say the three layers start out sized 8, 3, 5. The ideal allocation would be two llcs for the first one, one full llc for the third one and the rest in the remaining llc. Let's say the sizes shift and we end up with 5, 3, 8 and stabilizes. There is no way to achieve the ideal allocaiton without actively moving the allocation for the middle layer.

Even ignoring the need for dynamic reallocation, placing the layers at equidistances is unlikely to bring the allocation close to the ideal state given how they tend to differ widly in sizes. The big guys will overrun the positions for the smaller guys who then in turn will encroach into other layers' slots.

**Comment by hodgesds** [comment]
Yeah, having the ideal growth be something that is dynamic is definitely a better solution. This is just a short term hack to make it a little better.

**Pull request merged**